exercises/nondet_monad_cont/descr.md

```diff
@@ -0,0 +1,522 @@
+  the singleton sequence composed of just `v`.
+
+* The computation `m1 >>= m2` is the sequential composition of the
+  computations `m1` and `m2`. This composition operator is
+  asymmetric: whereas its first argument `m1` is a computation of
+  type `'a m`, its second argument `m2` is a function of type `a ->
+  'b m`. Every value `x` produced by `m1` is passed to `m2`,
+  yielding a computation `m2 x`. The sequence of values produced by
+  `m1 >>= m2` is the concatenation of the sequences of values
+  produced by the computations `m2 x`, where `x` ranges over the
+  values produced by `m1`.
+
+The constructor functions `fail` and `choose` are specific of the
+nondeterminism monad. `fail` can be thought of as a 0-ary
+disjunction, whereas `choose` is a binary disjunction. Together,
+they allow constructing the disjunction of an arbitrary number of
+computations.
+
+* The computation `fail` returns no result. In other words, it
+  produces an empty sequence of values.
+
+* The sequence of values produced by `choose m1 m2` is the
+  concatenation of the sequences of values produced by `m1`
+  and by `m2`.
+
+The constructor function `delay` is used to delay the construction
+of a computation until the moment when this computation must be
+executed. Indeed, a difficulty that arises in a strict programming
+language, such as OCaml, is that the arguments passed to constructor
+functions, such as `return` and `choose`, are evaluated immediately,
+at construction time. For instance, when one writes `choose e1 e2`,
+both of the OCaml expressions `e1` and `e2` are evaluated before
+`choose` is invoked. If the evaluation of `e2` performs nontrivial
+work, then this work arguably takes place too early: indeed, there
+should be no need to evaluate `e2` until all of the values produced
+by `e1` have been demanded. To remedy this, one may write `choose e1
+(delay (fun () -> e2))`. There, the expression `e2` is placed in the
+body of an anonymous function, so `e2` is not evaluated immediately.
+This anonymous function, whose type is `unit -> 'a m`,
+is converted by `delay` to a computation of type `'a m`.
+This conversion requires no serious work;
+it is performed in constant time.
+An intuitive reason why this is possible is that the type `'a m`
+represents a *suspended* computation already,
+so the type `unit -> 'a m` represents a *suspended suspended*
+computation, which is essentially the same thing; these types
+are interconvertible at no cost.
+
+The computations `e` and `delay (fun () -> e)` produce the same
+sequence of results. The only difference between them is the time at
+which the evaluation of `e` takes place: either immediately, or only
+when the first result is demanded.
+
+It is worth noting that in a lazy language, such as Haskell, there
+is no need for `delay`. In such a language, when one writes `choose
+e1 e2`, the expressions `e1` and `e2` are *not* evaluated
+immediately: they are evaluated only when their value is demanded.
```

Comment:
While "demanding a delayed value `v` of the form `fun () -> e`" is clearly an application `v ()`, a student may wonder how to express "demanding a value of a lazy expression `e`". I remember that it took me a little while to understand how this works in Haskell.

Reply:
(This is probably off-topic to explain that though.)